### PR TITLE
shake: embed data files into 'shake' executable

### LINKED ISSUE #268: Unable to build report from relocated shake executable
We build our shake executable on a build server and copy it to developer machines. This works, but running it work `--report` crashes with:

```
/some/path/on/build/server/shake-0.15.2/html/profile.html: openBinaryFile: does not exist (No such file or directory)
```

This is probably due to the use of cabal data dirs. I have no idea how to fix this, though.

### LINKED ISSUE #413: Option to embed data-files into the library
My team uses shake to build the product we develop. Me and one of my colleagues, we are maintaining the build system while others just use it. We would like it to be as independent as possible so we compile and distribute the build script as a static binary. Everything works great except reports - shake tries to find it's report templates in a compile-time-defined directory, which obviously does not exist when you run a static binary you've just downloaded. 
We would like to have a build option to embed the data files into binary so that the static binary could generate reports.
I have googled 'http://haddock.stackage.org/lts-5.0/file-embed-0.0.9.1/Data-FileEmbed.html', looks like it can be used to accompliish the task.
I think, I can do that by myself, but I'd like to discuss the feature beforehand.

### PR DESCRIPTION
**Pull request by thoughtpolice**
This patch directly embeds the set of needed re-distributable artifacts
in the Shake repository into the 'shake' executable by default, and,
optionally, the 'shake' library as well.

Without this change, Shake binaries cannot build HTML reports without
the Haskell toolchain installed. This means that users of Shake for e.g.
ninja compatibility are unable to build reports transparently.

There have been ideas floated around to embed data files at the end of
the binary, but this seems needlessly complicated. Instead, this patch
simply embeds the files in the binary at compile time if asked, using
Template Haskell.

One complicating factor that made this change hard in the past is that
GHC is moving to use Hadrian (and Shake) by default, and so adding a
stage2 feature dependency (TemplateHaskell requires GHCi) seriously
hurts the ability to bootstrap GHC: you would need an entire stage2
cross toolchain to build Hadrian, for instance.

However, there are some crucial distinctions to make here, and some
factors that make this change not so bad:

  - The `shake` executable, which provides Ninja compatibility, is NOT
  bound by the Hadrian/Haskell stage2 requirements. This means that even
  e.g.  CMake or external users of Shake can still benefit from such a
  change, if it is done carefully.

  - The `shake` library, which is what developers use to write build
  systems, MAY be bound by the stage2 restriction, i.e. cannot use
  Template Haskell, due to platform restrictions. But it also may not
  be; a user on x86_64-linux who is writing a custom build system may
  well want to deliver it to their developers, also running
  x86_64-linux.

  - Users of Haskell tooling, such as Stack or Cabal (new-build),
  generally have the ability to set per-package compilation flags for
  their projects. For instance, a single project using `cabal new-build`
  can globally enforce the usage of Shake, but only compiled with a
  specific set of build-time flags. While this wasn't always available,
  it is now generally widely available.

The third mitigating factor is what motivates this solution as the
simplest and easiest: to simply add support and help users in group 1
(currently myself), and also some of the users in group 2 by simply
requiring them to set flags for the dependent package.

In this setup, the `shake` library, by default, still DOES NOT use file
embedding. Rather, setting the flag `-fembed-files` at build time will
enable support for it in the library. Users of Shake who want to create
binaries that have profiling support "included" can simply enforce their
projects to have this flag enabled, and their binaries will always
support this feature.

The `shake` executable, however, ALWAYS has support for file-embedding
built in. This executable is generally useful only for Ninja
compatibility (or general demo purposes), so adding this requirement
does not seem especially onerous, and improves the lives of any user who
is currently depends on it.

This patch has been tested on a full CMake/Ninja build on a standalone
Ubuntu 16.04 machine, and was able to generate a working HTML build
report.

Closes #268. Closes #413.

### CODE REVIEW

**Comment by thoughtpolice** [comment]
@ndmitchell ping, is there anything you'd like to see in this PR for it to be merged, or do you fundamentally think it's a bad idea to go this way? I'd really like the default `shake` executable to support profiling ninja builds OOTB, at least...

**Comment by ndmitchell** [comment]
Very well argued and implemented. Something gives me hesitation, but perhaps its just a reflex fear from having been burnt by TemplateHaskell so many times. But I'm pretty convinced - can you fix up the conflict? (I would, but it seems I don't understand the reason for the change on that line, so don't know the why)

@snowleopard - can you confirm the above changes won't cause any `ghc` problems? It means `shake` the executable and test suite will require template-haskell`, but (by default) `shake` the library won't.

Again, many apologies for the delay in reviewing such useful functionality!

**Comment by snowleopard** [comment]
@ndmitchell As far as I understand, GHC won't be affected by this change since we only use the Shake library, not the executable.

**Comment by ndmitchell** [comment]
Thanks, I've pushed some subsequent patches cleaning a few bits up.

File: src/Development/Shake/Internal/Paths.hs [hunk 1]
```diff
@@ -51,13 +88,11 @@ getDataFile file = do
...
 hasDataFile file = anyM (\dir -> doesFileExist_ $ dir </> file) dataDirs
 
-
 readDataFileHTML :: FilePath -> IO LBS.ByteString
 readDataFileHTML file = LBS.readFile =<< getDataFile ("html" </> file)
 
-
 manualFiles :: [FilePath]
-manualFiles = map ("docs/manual" </>) ["Build.hs","main.c","constants.c","constants.h","build" <.> if isWindows then "bat" else "sh"]
+manualFiles = map ("docs/manual" </>) ["Build.hs","main.c","constants.c","constants.h","build.bat","build.sh"]
```
[the remark this answers is not in the document]
Comment (thoughtpolice) [reply src/Development/Shake/Internal/Paths.hs]: Oops, I think this was just a mistake on my part.

[the remark this answers is not in the document]
Comment (thoughtpolice) [reply src/Development/Shake/Internal/Paths.hs]: Looks good, thanks!

File: src/General/Template.hs [hunk 2]
```diff
@@ -13,13 +18,26 @@ import qualified Language.Javascript.DGTable as DGTable
 import qualified Language.Javascript.Flot as Flot
 import qualified Language.Javascript.JQuery as JQuery
 
+#ifdef FILE_EMBED
+import Data.FileEmbed
+import Language.Haskell.TH.Syntax ( runIO )
+#endif
 
+#ifdef FILE_EMBED
+libraries =
```
[the remark this answers is not in the document]
Comment (thoughtpolice) [reply src/General/Template.hs]: Looks good, thanks!

**Comment by thoughtpolice** [comment]
Thanks @ndmitchell!

**Pull request merged**